### PR TITLE
Only acquire semaphore lease if maxconnections is configured

### LINKED ISSUE #12459: Port forwarding doesn't work for kubernetes
Expected behaviour:
If I try to port-forward something with let's say `kubectl port-forward my-pod 8080` my local port 8080 should get forwarded to the port 8080 on that pod.

Current behaviour:
I get the error `error: error upgrading connection: missing parameter MaxLeases`. This is the case for port-forwarding to a pod or to a service.

Bug details:
- Teleport version
  - client: Teleport v9.2.0 git:v9.2.0-0-gc7db54496 go1.17.9
  - server/cluster inside kubernetes: Teleport v9.2.0 git:v9.2.0-0-gc7db54496 go1.17.9
- Recreation steps
  - Installing teleport with helm inside kubernetes
  - Log in on a client and try to port-forward something

Found out a bit later:
The same behaviour is if I try to create an interactive shell to a pod, like this: `kubectl exec -it my-pod -- bash`.

### PR DESCRIPTION
**Pull request by xacrimon**
Fixes #12459

### CODE REVIEW

**Review by r0mant — COMMENTED** [review]
@xacrimon In addition to testing manually, let's please add test coverage to make sure it does not regress again.

File: lib/kube/proxy/forwarder.go [hunk 1]
```diff
@@ -453,7 +453,14 @@ func (f *Forwarder) withAuth(handler handlerWithAuthFunc) httprouter.Handle {
...
 		}
-		if err := f.acquireConnectionLock(req.Context(), authContext); err != nil {
+
+		user := authContext.Identity.GetIdentity().Username
+		roles, err := getRolesByName(f, authContext.Identity.GetIdentity().Groups)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := f.AcquireConnectionLock(req.Context(), user, roles); err != nil {
```
Comment (r0mant) [inline lib/kube/proxy/forwarder.go:463]: Where is this semaphore released btw? I don't see if in the code unless I'm missing something.

Comment (xacrimon) [reply lib/kube/proxy/forwarder.go:463]: Ah, they're all cancelled when the `req.Context()` cancels. This happens whenever the request gets completed/the socket closes.

Comment (xacrimon) [reply lib/kube/proxy/forwarder.go:463]:
```
// AcquireSemaphoreLock attempts to acquire and hold a semaphore lease.  If successfully acquired,
// background keepalive processes are started and an associated lock handle is returned.  Cancelling
// the supplied context releases the semaphore.```
as per the semaphore godoc

**Review by r0mant — APPROVED** [review]

File: lib/kube/proxy/forwarder.go (hunk 1 above)
Comment (r0mant) [reply lib/kube/proxy/forwarder.go:463]: Got it!

**Review by rosstimothy — APPROVED** [review]

File: lib/kube/proxy/forwarder_test.go [hunk 2]
```diff
@@ -1053,3 +1055,92 @@ func (m *mockWatcher) Events() <-chan types.Event {
...
+			require.NoError(t, err)
+
+			sem := local.NewPresenceService(backend)
+			client := &mockSemaphoreClient{sem: sem}
+			forwarder := newTestForwarder(ctx, ForwarderConfig{
+				AuthClient: client,
+			})
+
+			for i := 0; i < testCase.connections; i++ {
+				err = forwarder.AcquireConnectionLock(ctx, user.GetName(), services.NewRoleSet(testCase.role))
```
Comment (zmb3) [inline lib/kube/proxy/forwarder_test.go:1140]: Are you intentionally ignoring all the last error here?

Comment (xacrimon) [reply lib/kube/proxy/forwarder_test.go:1140]: Yep. If this test is going to error, it's always going to error towards the tail end of the attempts we make make.

File: lib/kube/proxy/forwarder.go [hunk 3]
```diff
@@ -897,25 +904,23 @@ func wsProxy(wsSource *websocket.Conn, wsTarget *websocket.Conn) error {
...
-	if err != nil {
-		return trace.Wrap(err)
+func (f *Forwarder) AcquireConnectionLock(ctx context.Context, user string, roles services.RoleSet) error {
+	maxConnections := roles.MaxKubernetesConnections()
+	if maxConnections == 0 {
+		return nil
 	}
 
-	maxConnections := services.RoleSet(roles).MaxKubernetesConnections()
 	semLock, err := services.AcquireSemaphoreLock(ctx, services.SemaphoreLockConfig{
```
Comment (ibeckermayer) [inline lib/kube/proxy/forwarder.go:916]: Currently, `services.AcquireSemaphoreLock`'s comment string says "`If successfully acquired,
background keepalive processes are started and an associated lock handle is returned.`", however it appears that that isn't actually happening in that function, and it's the caller's responsibility to call
```
go semLock.KeepAlive(ctx)
```

We should either change the comment or the behavior.

**Review by ibeckermayer — APPROVED** [review]

File: lib/kube/proxy/forwarder_test.go (hunk 2 above)
Comment (zmb3) [reply lib/kube/proxy/forwarder_test.go:1140]: In that case I might suggest moving the error check inside the loop, with something like:

```
err = forwarder.AcquireConnectionLock(...)
if i == testCase.connections = 1 {
    testCase.assert(t, err_
}
```

This way it's clear that we only care about the error on the last invocation and doesn't look like a bug.

Comment (xacrimon) [reply lib/kube/proxy/forwarder_test.go:1140]: Ah, makes sense. Fixed

File: lib/kube/proxy/forwarder.go (hunk 3 above)
Comment (xacrimon) [reply lib/kube/proxy/forwarder.go]: Yeah, good point. I've amended `Acquire` to start the keepalive and made the keepalive function private.

**Pull request merged**